File: Documentation/5.4/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/HowTo/EntityChanges.cs
using Raven.Client.Documents;
using Raven.Client.Documents.Session;
using Raven.Documentation.Samples.Orders;
using Xunit;

namespace Raven.Documentation.Samples.ClientApi.Session.HowTo
{
    public class EntityChanges
    {
        public EntityChanges()
        {
            using (var store = new DocumentStore())
            {
                #region changes_1
                using (var session = store.OpenSession())
                {
                    // Store a new entity within the session
                    // =====================================
                    
                    Employee employee = new Employee {FirstName = "John", LastName = "Doe"};
                    session.Store(employee, "employees/1-A");
                    
                    // 'HasChanged' will be TRUE 
                    Assert.True(session.Advanced.HasChanged(employee));
                    
                    // 'HasChanged' will reset to FALSE after saving changes 
                    session.SaveChanges();
                    Assert.False(session.Advanced.HasChanged(employee));
                    
                    // Load & modify entity within the session
                    // =======================================
                    
                    employee = session.Load<Employee>("employees/1-A");
                    Assert.False(session.Advanced.HasChanged(employee)); // FALSE
                    
                    employee.LastName = "Brown";
                    Assert.True(session.Advanced.HasChanged(employee)); // TRUE
                    
                    session.SaveChanges();
                    Assert.False(session.Advanced.HasChanged(employee)); // FALSE
                }
                #endregion
                
                #region changes_2
                using (var session = store.OpenSession())
                {
                    // Store (add) a new entity, it will be tracked by the session
                    Employee employee = new Employee {FirstName = "John", LastName = "Doe"};
                    session.Store(employee, "employees/1-A");

                    // Get the changes for the entity in the session
                    // Call 'WhatChangedFor', pass the entity object in the param
                    DocumentsChanges[] changesForEmployee = session.Advanced.WhatChangedFor(employee);
                    Assert.Equal(changesForEmployee.Length, 1); // a single change for this entity (adding)
    
                    // Get the change type
                    DocumentsChanges.ChangeType changeType = changesForEmployee[0].Change;
                    Assert.Equal(changeType, DocumentsChanges.ChangeType.DocumentAdded);
    
                    session.SaveChanges();
                }
                #endregion
                
                #region changes_3
                using (var session = store.OpenSession())
                {
                    // Load the entity, it will be tracked by the session
                    Employee employee = session.Load<Employee>("employees/1-A");
                    
                    // Modify the entity
                    employee.FirstName = "Jim";
                    employee.LastName = "Brown";
                    
                    // Get the changes for the entity in the session
                    // Call 'WhatChangedFor', pass the entity object in the param
                    DocumentsChanges[] changesForEmployee = session.Advanced.WhatChangedFor(employee);
                   
                    Assert.Equal(changesForEmployee[0].FieldName, "FirstName");                           // Field name
                    Assert.Equal(changesForEmployee[0].FieldNewValue, "Jim");                             // New value
                    Assert.Equal(changesForEmployee[0].Change, DocumentsChanges.ChangeType.FieldChanged); // Change type
                    
                    Assert.Equal(changesForEmployee[1].FieldName, "LastName");
                    Assert.Equal(changesForEmployee[1].FieldNewValue, "Brown");
                    Assert.Equal(changesForEmployee[1].Change, DocumentsChanges.ChangeType.FieldChanged);
                    
                    session.SaveChanges();
                }
                #endregion
            }
        }
        
        private interface IFoo
        {
            #region syntax_1
            // HasChanged
            bool HasChanged(object entity);
            #endregion
        }
        
        private interface IFoo2
        {
            #region syntax_2
            // WhatChangedFor
            DocumentsChanges[] WhatChangedFor(object entity);
            #endregion
            
            #region syntax_3
            public class DocumentsChanges
            {
                public object FieldOldValue { get; set; }  // Previous field value
                public object FieldNewValue { get; set; }  // Current field value
                public ChangeType Change { get; set; }     // Type of change that occurred
                public string FieldName { get; set; }      // Name of field on which the change occurred
                public string FieldPath { get; set; }      // Path of field on which the change occurred
                public string FieldFullName { get; }       // Path + Name of field on which the change occurred
            }
            
            public enum ChangeType
            {
                DocumentDeleted,
                DocumentAdded,
                FieldChanged,
                NewField,
                RemovedField,
                ArrayValueChanged,
                ArrayValueAdded,
                ArrayValueRemoved
            }
            #endregion
        }
    }
}

File: Documentation/5.4/Samples/nodejs/client-api/session/howTo/entityChanges.js
import { DocumentStore } from "ravendb";

const documentStore = new DocumentStore();

async function example() {
    {
        //region changes_1
        const session = documentStore.openSession();
        
        // Store a new entity within the session
        // =====================================

        let employee = new Employee();
        employee.firstName = "John";
        employee.lastName = "Doe";
        await session.store(employee, "employees/1-A");

        // 'hasChanged' will be TRUE 
        assert.ok(session.advanced.hasChanged(employee));

        // 'hasChanged' will reset to FALSE after saving changes 
        await session.saveChanges();
        assert.ok(!session.advanced.hasChanged(employee));

        // Load & modify entity within the session
        // =======================================

        employee = await session.load("employees/1-A");
        assert.ok(!session.advanced.hasChanged(employee));   // FALSE

        employee.lastName = "Brown";
        assert.ok(session.advanced.hasChanged(employee));    // TRUE

        await session.saveChanges();
        assert.ok(!session.advanced.hasChanged(employee));   // FALSE
        //endregion
    }
    {
        //region changes_2
        const session = documentStore.openSession();
        
        // Store (add) a new entity, it will be tracked by the session
        let employee = new Employee();
        employee.firstName = "John";
        employee.lastName = "Doe";
        await session.store(employee, "employees/1-A");

        // Get the changes for the entity in the session
        // Call 'whatChangedFor', pass the entity object in the param
        const changesForEmployee = session.advanced.whatChangedFor(employee);
        
        // Assert there was a single change for this entity
        assert.equal(changesForEmployee.length, 1);

        // Get the change type
        const changeType = changesForEmployee[0].change;
        assert.equal(changeType, "DocumentAdded");
        
        await session.saveChanges();
        //endregion
    }
    {
        //region changes_3
        const session = documentStore.openSession();
        
        // Load the entity, it will be tracked by the session
        const employee = await session.load("employees/1-A");

        // Modify the entity
        employee.firstName = "Jim";
        employee.lastName = "Brown";

        // Get the changes for the entity in the session
        // Call 'whatChangedFor', pass the entity object in the param
        const changesForEmployee = session.advanced.whatChangedFor(employee);

        assert.equal(changesForEmployee[0].fieldName, "firstName");
        assert.equal(changesForEmployee[0].fieldNewValue, "Jim");
        assert.equal(changesForEmployee[0].change, "FieldChanged");

        assert.equal(changesForEmployee[1].fieldName, "lastName");
        assert.equal(changesForEmployee[1].fieldNewValue, "Brown");
        assert.equal(changesForEmployee[1].change, "FieldChanged");
        //endregion
    }
}

{
    let entity;
    
    //region syntax_1
    session.advanced.hasChanged(entity);
    //endregion
    
    //region syntax_2
    session.advanced.whatChangedFor(entity);
    //endregion

    //region syntax_3
    class DocumentsChanges {
        // Previous field value
        fieldOldValue; // object

        // Current field value
        fieldNewValue; // object        
        
        // Name of field on which the change occurred
        fieldName; // string

        // Path of field on which the change occurred
        fieldPath; // string

        // Path + Name of field on which the change occurred
        fieldFullName; // string

        // Type of change that occurred, can be: 
        //     "DocumentDeleted"
        //     "DocumentAdded"
        //     "FieldChanged"
        //     "NewField"
        //     "RemovedField"
        //     "ArrayValueChanged"
        //     "ArrayValueAdded"
        //     "ArrayValueRemoved"
        change; // string
    }
    //endregion
}
